### ISSUE TITLE
Java 10/11 should be supported

### ISSUE THREAD

**Issue by ajrepp** [bazelbuild/bazel#5723]
### Description of the problem / feature request:

It seems like Bazel doesn't support building projects with Java 10 or Java 11. Java 10 is scheduled to be deprecated when Java 11 is released, currently estimated for September 25, 2018. In order to test and update our code base for Java 11, we need to be able to build it with Bazel.

### Feature requests: what underlying problem are you trying to solve with this feature?

Bazel should support building against Java 10/11.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Follow the steps at https://github.com/ajrepp/java-upgrade-toy. The output is:
```
$ bazel build //:x --javabase=//:jdk10-jdk --java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 --host_javabase=//:jdk10-jdk
INFO: Build options have changed, discarding analysis cache.
INFO: Analysed target //:x (0 packages loaded).
INFO: Found 1 target...
ERROR: /Users/andrew.repp/code/java-upgrade-toy/BUILD:1:1: Building x.jar (1 source file) failed (Exit 1)
java.lang.IllegalArgumentException: release version 10 not supported
	at jdk.compiler/com.sun.tools.javac.main.Arguments.error(Arguments.java:917)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.handleReleaseOptions(Arguments.java:310)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.processArgs(Arguments.java:370)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.init(Arguments.java:246)
	at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:185)
	at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:97)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$2.invokeJavac(SimpleJavaLibraryBuilder.java:122)
	at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:125)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:133)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
Target //:x failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1.190s, Critical Path: 0.94s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

### What operating system are you running Bazel on?

macOS X 10.13.6

### What's the output of `bazel info release`?

development version

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

bazel build //scripts/packages

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

https://github.com/bazelbuild/bazel.git
0a635c5236ce30ea84b765ce752267992733a649
0a635c5236ce30ea84b765ce752267992733a649

###  Have you found anything relevant by searching the web?

- https://groups.google.com/forum/#!topic/bazel-discuss/quLTk1m4JE8
- https://groups.google.com/forum/#!topic/bazel-discuss/YR7e2wDDjzg
- #3410 
- #4812
- https://github.com/bazelbuild/bazel/commit/4c9149d558161e7d3e363fb697f5852bc5742a36#diff-d2a96a9fef0d2a072ca555fb0fa4a80c

### Any other information, logs, or outputs that you want to share?

We haven't been able to find a definitive answer anywhere about new Java version support or easy-to-understand documentation for how to use new versions. Any information would be appreciated!

**Comment by ajrepp** [comment]
It looks like this might be related to https://github.com/bazelbuild/bazel/blob/master/tools/jdk/default_java_toolchain.bzl#L37-L38 -- those jars might not support Java 10+?

**Comment by ajrepp** [comment]
@cushon is Java 11 supported as well? Java 10 will no longer receive support once Java 11 is released, which is [scheduled for September 25th](http://openjdk.java.net/projects/jdk/11/) (less than 5 weeks away!).

**Comment by cushon** [comment]
Note that using JDK 11 as a `--javabase` should already work. The part that requires changes is supporting the Java 11 language level (`--release 11`, etc.), I will defer to @lberki for that.

**Comment by davido** [comment]
> Note that using JDK 11 as a --javabase should already work.

How would I set-up local jdk 11 to be used as `--javabase` to build gerrit with local jdk 11? 
```
  $ echo $JAVA_HOME
  /usr/lib64/jvm/java-11
```

As you pointed out in this comment: [1], there are three JDKs used in Bazel, I think we want to change 2) and 3).

So I define this `java_runtime` rule:

```
java_runtime(
    name = "jdk11",
    java_home = "/usr/lib64/jvm/java-11",
    visibility = ["//visibility:public"],
)
```

And using it like this:

```
 $ bazel build --javabase //:jdk11 --host_javabase //:jdk11 java/com/google/gerrit/server:server
```

That works, and still raises the question how to pass post Java arguments to the tests, like we do for Java 9 and Java 10:

```
  $ bazel build --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 --java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 :release
```

However, I would expect that the class version produced by the command above is major version 55. To my surprise, the major version is 52 (Java 8):

```
$ /usr/lib64/jvm/java-11/bin/javap -v -classpath bazel-bin/java/com/google/gerrit/server/libserver.jar com.google.gerrit.server.util.SocketUtil | more
Classfile jar:file:///home/davido/projects/gerrit2/bazel-bin/java/com/google/gerrit/server/libserver.jar!/com/google/gerrit/server/util/SocketUtil.class
  Last modified Jan 1, 1980; size 3601 bytes
  MD5 checksum 4379cabbb77f09e7317b9d14f4e6336b
  Compiled from "SocketUtil.java"
public final class com.google.gerrit.server.util.SocketUtil
  minor version: 0
  major version: 52
```

Am I missing something? Are we passing somewhere in the Bazel toolchain hard coded Java target = 8?

* [1] https://github.com/bazelbuild/bazel/issues/5594#issuecomment-407169981

**Comment by davido** [comment]
If I pass, `--release 11` in my previous command, that I'm, getting error: `release version 11 not supported`. Can it be, that the JDK 11 doesn't actually used?

```
$ bazel build --javabase //:jdk11 --host_javabase //:jdk11 --javacopt="--release 11" //java/com/google/gerrit/server:server
INFO: Build options have changed, discarding analysis cache.
INFO: Analysed target //java/com/google/gerrit/server:server (1 packages loaded).
INFO: Found 1 target...
INFO: From Executing genrule @bazel_tools//tools/jdk:platformclasspath [for host]:
warning: [options] bootstrap class path not set in conjunction with -source 8
1 warning
ERROR: /home/davido/projects/gerrit2/java/org/eclipse/jgit/BUILD:37:1: Building java/org/eclipse/jgit/libserver.jar (2 source files) failed (Exit 1)
java.lang.IllegalArgumentException: release version 11 not supported
	at jdk.compiler/com.sun.tools.javac.main.Arguments.error(Arguments.java:897)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.handleReleaseOptions(Arguments.java:311)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.processArgs(Arguments.java:350)
	at jdk.compiler/com.sun.tools.javac.main.Arguments.init(Arguments.java:246)
	at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:185)
	at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:97)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$1.invokeJavac(SimpleJavaLibraryBuilder.java:106)
	at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:109)
	at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:117)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
	at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
Target //java/com/google/gerrit/server:server failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 6.300s, Critical Path: 5.86s
INFO: 3 processes: 1 remote cache hit, 2 worker.
FAILED: Build did NOT complete successfully
```

**Comment by cushon** [comment]
> I think we want to change [--host_javabase] and [--javabase].

`--host_javabase=11` isn't tested or supported yet. It probably works, but the recommended configuration today is to use the default embedded JDK 10 as the `--host_javabase`. (That may [be changing](https://groups.google.com/d/msg/bazel-sig-jvm/wA9f-B6Tl5w/SnxzztZEAQAJ); we'll see.)

The part I expected to work was specifically using JDK 11 as a `--javabase`.

> However, I would expect that the class version produced by the command above is major version 55. To my surprise, the major version is 52 (Java 8):

I can't reproduce. With gerrit @ 8f26af11797075aab14d9c17d1e69a30e00aa9e5:

```
$ bazel version
Build label: 0.17.0rc1
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Aug 22 18:43:25 2018 (1534963405)
Build timestamp: 1534963405
Build timestamp as int: 1534963405
$ bazel build --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 --java_toolchain=@bazel_tools//tools/jdk:toolchain_java10 //java/com/google/gerrit/server:libserver.jar
...
Target //java/com/google/gerrit/server:libserver.jar up-to-date:
  bazel-bin/java/com/google/gerrit/server/libserver.jar
$ javap -v -cp bazel-bin/java/com/google/gerrit/server/libs
erver.jar com.google.gerrit.server.util.SocketUtil | head
Classfile jar:file:///usr/local/google/home/cushon/src/gerrit/bazel-bin/java/com/google/gerrit/server/libserver.jar!/com/google/gerrit/server/util/SocketUtil.class
  Last modified Jan 1, 1980; size 3601 bytes
  MD5 checksum f62f6ea242e64c848c0f6723ff47e57a
  Compiled from "SocketUtil.java"
public final class com.google.gerrit.server.util.SocketUtil
  minor version: 0
  major version: 54
```

**Comment by cushon** [comment]
> If I pass, `--release 11` in my previous command, that I'm, getting `error: release version 11 not supported`. Can it be, that the JDK 11 doesn't actually used?

`--release` is a javac flag, javac is configured separately from the `javabase`s as part of the `java_toolchain`, and the toolchain that ships with Bazel is using a JDK 10 javac. It will need to be updated to a JDK 11 javac to support `--release 11`. That's partly what this bug is tracking.

**Comment by cushon** [comment]
> How can I produce today major version: 55 with Bazel@HEAD?

That isn't currently supported with the default toolchain: it's is using a JDK 10 javac, which will need to be updated to a JDK 11 javac to support --release 11. Again, that is part of the work being tracked by this bug.

> `--javabase //:jdk11 --host_javabase //:jdk11` doesn't work and produces major version 52

That isn't supposed to produce v52 class files: the language level is configured by `--java_toolchain`, not `--host_javabase` or `--javabase`.

**Comment by davido** [comment]
@cushon 

> That isn't currently supported with the default toolchain: it's is using a JDK 10 javac, which will need to be updated to a JDK 11 javac to support --release 11. Again, that is part of the work being tracked by this bug.

I see, thanks for clarifying. I see 2 fundamental problems with thsi approach:

# Waiting indefinitely for Bazel to support new and shiny JDK versions
# During upgrade to latest and greatest JDK in bazel as default toolchain, we see breakages supporting older, but still new enough JDK versions, as discovered here: [1]. Yes, it is Java bug, but that doesn't change the fact, that there currently no way to build Gerrit with JDK9 on Bazel@HEAD

Can't language level be configured by --host_javabase or --javabase instead of --java_toolchain? Or can --java_toolchain and --host_java_toolchain be easily used for custom JDKs?

Do you have an working example how could i set-up custom toolchain with JDK11 to produce gerrit with --release 11 (major veersion 55) with Bazel@HEAD?

**Comment by cushon** [comment]
> there currently no way to build Gerrit with JDK9 on Bazel@HEAD

To confirm: do you want to build with the Java 9 language level and target JDK 9? Do you care about the `--host_javabase` version as long as it supports that language level?

> Can't language level be configured by `--host_javabase` or

If you want to use whatever the latest supported language level is from the `--host_javabase`, you could use VanillaJavaBuilder (which should work with most host JDKs) and leave the source/target/bootclasspath unset so they default to the latest supported:

```bzl
default_java_toolchain(
    name = "toolchain_vanilla",
    forcibly_disable_header_compilation = True,
    javabuilder = [":vanillajavabuilder"],
    jvm_opts = [],
)
```

e.g.

```shell
for v in $(seq 8 11); do
  bazel build \
    --host_javabase=:jdk${v} \
    --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla \
    --java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla \
     //java/com/google/gerrit/server:libserver.jar
  javap -v -cp bazel-bin/java/com/google/gerrit/server/libserver.jar \
    com.google.gerrit.server.util.SocketUtil | grep "major version"
done
```

```
  major version: 52
  major version: 53
  major version: 54
  major version: 55
```

**Comment by davido** [comment]
> To confirm: do you want to build with the Java 9 language level and target JDK 9?

Yes.

> Do you care about the --host_javabase version as long as it supports that language level?

No.

> If you want to use whatever the latest supported language level is from the --host_javabase, you could use VanillaJavaBuilder [...]

In which case I wouldn't be able to use error prone and strict deps, and that's unfortunate.

**Comment by cushon** [comment]
re: error prone and strict deps--I'm glad those features are useful, you may want to read [this thread](https://groups.google.com/d/topic/bazel-sig-jvm/wA9f-B6Tl5w/discussion) if you haven't already and share any feedback you have.

I had an idea for working around the bug with JDK 10 and `--release 9`: https://github.com/bazelbuild/bazel/pull/5982

With that patch, the following works:

```bash
for v in $(seq 8 10); do
  bazel build --verbose_failures \
    --java_toolchain=@bazel_tools//tools/jdk:toolchain_java${v} \
    //java/com/google/gerrit/server:libserver.jar 
  javap -v -cp bazel-bin/java/com/google/gerrit/server/libserver.jar com.google.gerrit.server.util.SocketUtil | grep "major version"
done
```

```
  major version: 52
  major version: 53
  major version: 54
```

**Comment by davido** [comment]
> I had an idea for working around the bug with JDK 10 and --release 9: #5982 

Thanks!

I tried to use vanilla `java_toolchain` and it feels, like it could be simplify and naively supported by Bazel: #5984. WDYT?

Meantime, I added vanilla `java_toolchain` to Gerrit build tool chain: [1] to be able to build with upcoming JDK 11. It turns out, that the build is failing in rules_closure: [2] and all tests are failing too. I'm not sure yet, why the acceptance tests are failing, all I see is this:

```
 $ bazel test -s --host_javabase=:jdk11 --host_java_toolchain=//:toolchain_vanilla --java_toolchain=//:toolchain_vanilla //javatests/com/google/gerrit/acceptance/ssh:ssh
[...]
  $ cat /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/execroot/gerrit/bazel-out/k8-fastbuild/testlogs/javatests/com/google/gerrit/acceptance/ssh/ssh/test.log
There were 36 failures:
1) withoutMessage(com.google.gerrit.acceptance.ssh.AbandonRestoreIT)

BazelTestRunner exiting with a return value of 1
```

It seems that the JUnit-Log is not complete, may be there is some Bazel/JUnit integration issues on JDK11 (and would worth its own issue)?

* [1] https://gerrit-review.googlesource.com/c/gerrit/+/194040
* [2]

**Comment by cushon** [comment]
> It seems that the JUnit-Log is not complete may be there is some Bazel/JUnit integration issues on JDK11 (and would worth its own issue)?

If you look at the test log (or use `--test_summary=detailed`) it's a guice/ASM issue:

```
1) Error injecting method, com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: java.lang.UnsupportedOperationException
  at com.google.inject.assistedinject.FactoryProvider2.initialize(FactoryProvider2.java:716)
  at com.google.inject.assistedinject.FactoryModuleBuilder$1.configure(FactoryModuleBuilder.java:317) (via modules: com.google.gerrit.sshd.SshModule -> com.google.inject.assistedinject.FactoryModuleBuilder$1)
...
Caused by: java.lang.UnsupportedOperationException
        at org.objectweb.asm.ClassVisitor.visitNestHostExperimental(ClassVisitor.java:158)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:541)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:391)
        at com.google.inject.internal.util.LineNumbers.<init>(LineNumbers.java:64)
```

guice is using the `ASM6` API level, and the nest-related attributes that were added in 11 are only supported with `ASM7_EXPERIMENTAL`: https://gitlab.ow2.org/asm/asm/blob/master/asm/src/main/java/org/objectweb/asm/ClassVisitor.java#L158

But in general, sure: any time you have a minimal repro that demonstrates there's a bug in Bazel, please report it.

**Comment by davido** [comment]
> If you look at the test log (or use --test_summary=detailed) it's a guice/ASM issue:[...]

Thanks for pointing this out. Somehow the stack trace was reported on JDK10 for the same guice/ASM issue without providing this option: `--test_summary=detailed`. May be this is related to JavaBuilder vs. VanillaJavaBuilder?

Anyway, I fixed that issue in this PR: [1], conducted this guice release: [2] with new artifacts and consumed them in this CL: [3]. Now gerrit can be built on JDK11.

Thanks for your help and big thank to the whole Bazel team! I think it's very impressive that such complex project as Gerrit (from build tool chain point of view) can be built on all current released JDKs: 8,9,10 and even on the upcoming JDK11 ;-).

[1] https://github.com/google/guice/pull/1203
[2] https://github.com/davido/guice/releases/tag/4.2.1
[3] https://gerrit-review.googlesource.com/c/gerrit/+/194040

**Comment by davido** [comment]
> The choice of JavaBuilder implementation shouldn't affect the way test output is handled, and I can't observe a difference between the test output with JDK 10 and 11. Can you share the repro for that?

Yes, you are right. I cannot reproduce this any more. I tried on JDK 11 and reverted my Guice/ASM fix and now I see, in fact even without `--test_summary=detailed` all failure stacks for all tests are there: [1]. Something was wrong on my side.

* [1] http://paste.openstack.org/show/728804/

**Comment by racekiller** [comment]
DO i have to uninstall JDK10 and install JDK 9 then? I am trying to install bazel 0.4.5 and I am getting this  error

`Problem with java installation: couldn't find/access rt.jar in /Library/Java/JavaVirtualMachines/jdk-10.0.2.jdk/Contents/Home`

is it possible that anyone tell mw what to do? I need to finish the GPU CUDA installation and I have macOS 10.13

thanks

**Comment by davido** [comment]
> I am trying to install bazel 0.4.5 and I am getting this error

Why are you using such an ancient Bazel version? Current Bazel version is 0.16.1.

**Comment by cushon** [comment]
@racekiller, can you try upgrading to a newer version of Bazel? If that isn't an option or it doesn't help with the problem you're seeing, please file a new bug and include the steps to reproduce the problem you're seeing.

**Comment by davido** [comment]
@cushon I have re-tried to build Gerrit on master with this series applied: [1]. I used Bazel@HEAD with your [CL](https://bazel-review.googlesource.com/c/bazel/+/74550). Gerrit build is failing on Java 11 here now (and `java/com/google/gerrit/reviewdb/client/Comment.java` wasn't changed recently):

```
$ b9 build --host_javabase=:absolute_javabase --define=ABSOLUTE_JAVABASE=/usr/lib64/jvm/java-11 --host_java_toolchain=//:toolchain_vanilla --java_toolchain=//:toolchain_vanilla :headless
Starting local Bazel server and connecting to it...
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:98:7: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:99:7: Make sure that you are running at least Bazel 0.14.0.
INFO: Analysed target //:headless (221 packages loaded).
INFO: Found 1 target...
[...]
external/com_google_protobuf/src/google/protobuf/descriptor.h:1282:26: note: 'class google::protobuf::FileDescriptor' declared here
 class LIBPROTOBUF_EXPORT FileDescriptor {
                          ^~~~~~~~~~~~~~
ERROR: /home/davido/projects/gerrit2/java/com/google/gerrit/reviewdb/BUILD:7:1: Building java/com/google/gerrit/reviewdb/libclient-class.jar (32 source files) failed (Exit 1)
java/com/google/gerrit/reviewdb/client/Comment.java:261: error: cannot access com.google.gerrit.extensions.client.Comment.Range
  bad class file: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/execroot/gerrit/bazel-out/k8-fastbuild/bin/java/com/google/gerrit/extensions/libclient-ijar.jar(/com/google/gerrit/extensions/client/Comment$Range.class)
    class file is invalid for class com.google.gerrit.extensions.client.Comment$Range
    Please remove or make sure it appears in the correct subdirectory of the classpath.
java/com/google/gerrit/reviewdb/client/Comment.java:150: error: cannot find symbol
  symbol:   class Range
  location: class com.google.gerrit.extensions.client.Comment
java/com/google/gerrit/reviewdb/client/PatchLineComment.java:17: error: cannot find symbol
  symbol:   class Range
  location: class com.google.gerrit.extensions.client.Comment
java/com/google/gerrit/reviewdb/client/PatchLineComment.java:282: error: cannot find symbol
  symbol:   class Range
  location: class com.google.gerrit.reviewdb.client.PatchLineComment
java/com/google/gerrit/reviewdb/client/Comment.java:236: error: reference to Range is ambiguous
  both constructor Range(com.google.gerrit.reviewdb.client.Comment.Range) in com.google.gerrit.reviewdb.client.Comment.Range and constructor Range(Range) in com.google.gerrit.reviewdb.client.Comment.Range match
Target //:headless failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 70.316s, Critical Path: 14.91s
INFO: 247 processes: 238 linux-sandbox, 9 worker.
FAILED: Build did NOT complete successfully
```

Any clue why are we seeing this error now, and why it worked before (as reported earlier in this thread)?

[1] https://gerrit-review.googlesource.com/c/gerrit/+/194040

**Comment by cushon** [comment]
> `libclient-ijar.jar ... class file is invalid`
> Any clue why are we seeing this error now, and why it worked before (as reported earlier in this thread)?

Thanks for reporting that: I don't think it was working properly before, we might have just been avoiding the bug you noticed. The JDK 11 support that was added to ijar is half-baked: https://github.com/bazelbuild/bazel/issues/5987#issuecomment-420538341

**Comment by cushon** [comment]
It's blocked on upgrading the host javabase, which is blocked on getting`--incompatible_use_jdk10_as_host_javabase` released so we can flip the default.

I think it's also blocked on #6316, which will enable upgrading to a javac that doesn't support JDK 8 without regressing `--host_javabase=<jdk8>` support (i.e. providing a way to use non-VanillaJavaBuilder).

@lberki am I forgetting anything?

**Comment by lberki** [comment]
I think that's it.

**Comment by georgewfraser** [comment]
Is there any documentation of how to use Bazel with Java 10? The [tutorial](https://docs.bazel.build/versions/master/tutorial/java.html) seems to describe Java 8, it's not obvious from the [user manual](https://docs.bazel.build/versions/master/user-manual.html) what combination of arguments might lead to Java 10 working. When I did 

```bash
bazel build --javacopt="--release 10" {target}
```

I got the error:

```
java.lang.IllegalArgumentException: release version 10 not supported
        at jdk.compiler/com.sun.tools.javac.main.Arguments.error(Arguments.java:917)
        at jdk.compiler/com.sun.tools.javac.main.Arguments.handleReleaseOptions(Arguments.java:310)
        at jdk.compiler/com.sun.tools.javac.main.Arguments.processArgs(Arguments.java:370)
        at jdk.compiler/com.sun.tools.javac.main.Arguments.init(Arguments.java:246)
        at jdk.compiler/com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:185)
        at com.google.devtools.build.buildjar.javac.BlazeJavacMain.compile(BlazeJavacMain.java:97)
        at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder$1.invokeJavac(SimpleJavaLibraryBuilder.java:106)
        at com.google.devtools.build.buildjar.ReducedClasspathJavaLibraryBuilder.compileSources(ReducedClasspathJavaLibraryBuilder.java:54)
        at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.compileJavaLibrary(SimpleJavaLibraryBuilder.java:109)
        at com.google.devtools.build.buildjar.SimpleJavaLibraryBuilder.run(SimpleJavaLibraryBuilder.java:117)
        at com.google.devtools.build.buildjar.BazelJavaBuilder.processRequest(BazelJavaBuilder.java:105)
        at com.google.devtools.build.buildjar.BazelJavaBuilder.runPersistentWorker(BazelJavaBuilder.java:67)
        at com.google.devtools.build.buildjar.BazelJavaBuilder.main(BazelJavaBuilder.java:45)
```

It would be nice if there were basic instructions for how to build "hello world" with Java 10.

**Comment by davido** [comment]
You can configure Bazel build with Java 10 and 11. You have to use VanillaJavaBuilder for that. On Bazel master it can be easily done, because vanilla_toolchain was added to Bazel. On recent Bazel releases, it's a bit verbose to achieve, as vanilla_toolchain has to be added, but something like that should work (see working example in [Gerrit Code Review project](https://gerrit.googlesource.com/gerrit)):

```python
load(
    "@bazel_tools//tools/jdk:default_java_toolchain.bzl",
    "default_java_toolchain",
)

config_setting(
    name = "java_next",
    values = {
        "java_toolchain": ":toolchain_vanilla",
    },
)

java_runtime(
    name = "absolute_javabase",
    java_home = "$(ABSOLUTE_JAVABASE)",
    visibility = ["//visibility:public"],
)

# TODO(davido): Switch to consuming it from @bazel_tool//tools/jdk:toolchain_vanilla
# when my change is included in released Bazel version:
# https://github.com/bazelbuild/bazel/commit/0bef68e054eccecd690e5d9f46db8a0c4b2d887a
default_java_toolchain(
    name = "toolchain_vanilla",
    forcibly_disable_header_compilation = True,
    javabuilder = ["@bazel_tools//tools/jdk:VanillaJavaBuilder_deploy.jar"],
    jvm_opts = [],
)
```

Absolute javabase indirection is needed to keep non portable location of Java 10 (and newer JDK versions) out of BUILD files.

`config_setting` is added to allow to make conditional action in you build. We use it JGit, for example, to add additional dependencies, that were removed in recent JDK versions, e.g.:

```
POST_JAVA9_DEPS = ["//lib:jaxb-api"]

java_library(
    name = "jgit",
    srcs = SRCS,
    resource_strip_prefix = "org.eclipse.jgit/resources",
    resources = RESOURCES,
    deps = [
        "//lib:javaewah",
        [...]
    ] + select({
        "//:java_next": POST_JAVA9_DEPS,
        "//conditions:default": [],
    }),
)
```

To invoke with Bazel vanilla_tollchain:

```
  $ bazel build \
        --host_javabase=:absolute_javabase \
        --define=ABSOLUTE_JAVABASE=/usr/lib64/jvm/java-10 \
        --host_java_toolchain=//:toolchain_vanilla \
        --java_toolchain=//:toolchain_vanilla \
        //...
```

Note 1: You could use the same approach to support newer JDKs (and we do indeed), just replace in the command above java-10 with java-11 or java-12, ...

Note 2: To tun tests, additional argument must be passed: `--javabase=:absolute_javabase`, or alternatively, JAVA_HOME should be set to the JDK used, a lá:

```
  $ JAVA_HOME=/usr/lib64/jvm/java-10 bazel test [...]
```

**Comment by cushon** [comment]
> `# TODO(davido): Switch to consuming it from @bazel_tool//tools/jdk:toolchain_vanilla`

`@bazel_tool//tools/jdk:toolchain_vanilla` made it into 0.19x.

Also keep in mind that approach is unsupported; this bug is tracking full support:

https://github.com/bazelbuild/bazel/blob/f8be43cebfe5d9ee80e838c883fb305969ece743/tools/jdk/BUILD#L227-L233

**Comment by davido** [comment]
> @bazel_tool//tools/jdk:toolchain_vanilla made it into 0.19x.

Thanks for pointing this out. I confirmed, that it works as expected in 0.19.0, so that I was able to remove both `toolchain_vanilla` and `absolute_javabase` and this invocation just works:

```
bazel build \
    --define=ABSOLUTE_JAVABASE=/usr/lib64/jvm/java-11 \
    --host_javabase=@bazel_tools//tools/jdk:absolute_javabase \
    --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla \
    --java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla \
    :release
```

> Also keep in mind that approach is unsupported; this bug is tracking full support: [...]

This is well known and understood. Bazel already support with JavaBuilder Java 9:

```
  $ bazel build \
      --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java9 \
      --java_toolchain=@bazel_tools//tools/jdk:toolchain_java9 \
      :release
```

My plan is to adjust the instruction and move from VanillaJavaBuilder to JavaBuilder that would remove those restriction that you've mentioned, for newer Java versions, e.g.:

```
  $ bazel build \
      --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_java11 \
      --java_toolchain=@bazel_tools//tools/jdk:toolchain_java11 \
      :release
```

**Comment by wcurrie** [comment]
One (minor?) problem with the setup @uri-canva shows above. When compiling using `java_hostbase` pointed to jdk10 (or bazel packaged with nix to use jdk10) external libraries are missing from the `.jdeps` files written by `StrictJavaDepsPlugin` and `DependencyModule`.  The only real problem this causes is the intellij bazel plugin fails to import external libraries (for modules that aren't in the working set). We have `--strict_java_deps=OFF`.

From debugging it looks like with jdk10 as `java_hostbase` during compilation https://github.com/bazelbuild/bazel/blob/0b2352de3101e87647d083f6089246079dda0f75/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/plugins/dependency/ImplicitDependencyExtractor.java#L125 returns absolute `Path`s instead of relative ones. This breaks `DependencyModule.buildDependenciesProto`.

Example https://github.com/wcurrie/examples/commit/13f8b7e35702452a12bda6190bceadf0811bb300 based off  java-maven in https://github.com/bazelbuild/examples. With jdk8 as `java_hostbase` guava is included in the .jdeps output. With jdk10 it's missing.

Maybe it's more practical to work around this issue at the intellij plugin level. We've done this internally.

With this hack of `DependencyModule` (https://github.com/wcurrie/bazel/commit/452042c4e7170618fc2a0553317b7bfad4cc0c60) I could get .jdeps files including external libraries and jdk 10 support.

**Comment by cushon** [comment]
> absolute `Paths` instead of relative ones. This breaks `DependencyModule.buildDependenciesProto`.

The default javac has a fix for that issue:

https://github.com/google/error-prone-javac/commit/9901a45ff63c1d7bc9263007147dc866272b7fdc

That's one of the reasons Strict Deps doesn't work with the 'vanilla' toolchain:

https://github.com/bazelbuild/bazel/blob/44cadaf0becf43062a5d453208f8ecc741ec6e4a/tools/jdk/BUILD#L212-L214

**Comment by cushon** [comment]
> make its way into the javac bundled with bazel?

Yep. However it's blocked on #6316. The JDK 11 javac needs a JDK 11 `--host_javabase`, so in order to provide a JDK 11 javac without regressing support for earlier `--host_javabase` versions, we're planning to decouple the Java tools (including javac) from the Bazel distribution so you can select the appropriate version.

**Comment by benjaminp** [comment]
You can almost get Bazel itself to build wtih Java 11 using the vanilla toolchain. However, it fails because the vendored jarjar uses an old version of asm. Is there a plan update that?

**Comment by cushon** [comment]
@benjaminp Bazel at head is using ASM 7: https://github.com/bazelbuild/bazel/commit/f4a35eb934dc4a6ac3439830c1d09c7794d14496

**Comment by benjaminp** [comment]
Right, the problem is jarjar:
```
$ bazel build --stamp src:bazel_nojdk --java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla
ERROR: src/java_tools/junitrunner/java/com/google/testing/coverage/BUILD:54:1: Executing genrule //src/java_tools/junitrunner/java/com/google/testing/coverage:Jacoco_jarjar failed (Exit 1) bash failed: error executing command /bin/bash -c ... (remaining 1 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
Exception in thread "main" java.lang.UnsupportedOperationException: This feature requires ASM7
	at org.objectweb.asm.ClassVisitor.visitNestHost(ClassVisitor.java:150)
	at org.objectweb.asm.ClassVisitor.visitNestHost(ClassVisitor.java:153)
	at org.objectweb.asm.commons.ClassRemapper.visitNestHost(ClassRemapper.java:185)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:541)
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:391)
	at com.tonicsystems.jarjar.util.JarTransformer.process(JarTransformer.java:34)
	at com.tonicsystems.jarjar.util.JarProcessorChain.process(JarProcessorChain.java:36)
	at com.tonicsystems.jarjar.MainProcessor.process(MainProcessor.java:96)
	at com.tonicsystems.jarjar.util.StandaloneJarProcessor.run(StandaloneJarProcessor.java:44)
	at com.tonicsystems.jarjar.Main.process(Main.java:93)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.tonicsystems.jarjar.MainUtil.runMain(MainUtil.java:36)
	at com.tonicsystems.jarjar.Main.main(Main.java:49)
Target //src:bazel_nojdk failed to build
```

**Comment by cushon** [comment]
That error message is referring to the API level, not the library version: #6692

**Comment by uri-canva** [comment]
Now that Oracle Java 8 is past EOL is there any update on this? I see there's quite a bit of progress on #6316, but is that the only thing that is blocking Java 11 support? Can we use Java 11 with `--incompatible_use_remote_java_toolchain`?

**Comment by lberki** [comment]
That's the state of the affairs, I think. The flag flip for #6316 should happen pretty soon (/cc @iirina ).

**Comment by tomikazi** [comment]
Does this mean 0.23.0? Our project community is eagerly waiting for being able to build against JDK11 using off-the-shelf Bazel.

**Comment by lberki** [comment]
0.24, unfortunately. The flag is not flipped in 0.23 by default :(

**Comment by cushon** [comment]
I think #6316 is necessary but not sufficient: it makes it possible to add support for the Java 11 language level to the toolchain without regressing JDK 8 host_javabase support (because we can provide different versions of the tools). But the work to add support for 11 hasn't happened yet, and I'm not sure it's in scope for #6316.

**Comment by softprops** [comment]
Just noting java 12 is here https://hackernoon.com/announcement-java12-has-released-435e659affee

**Comment by greggdonovan** [comment]
If built-in support for Java 11/12 is going to take more time, documentation on how to enable it yourself with toolchains and/or other configuration would be very helpful.

**Comment by davido** [comment]
@greggdonovan It is already possible to support Java 11/12/... with recent Bazel releases with `toolchain_vanilla`. See my coment above: https://github.com/bazelbuild/bazel/issues/5723#issuecomment-435571763

This feature is also documented in Gerrit Code Review build toolchain documentation: [1] and there is even CI job in Bazel@CI for that.

[1] https://github.com/GerritCodeReview/gerrit/blob/master/Documentation/dev-bazel.txt#L17-L67
